File: challs/misconfiguration.dir/misconfiguration.pl
#! /usr/bin/env perl

use strict;
use warnings;
use DBI;
use Digest::SHA  qw(sha1_hex);
use MIME::Base64;
use Mojolicious;
use Mojo::Log;
use Mojo::URL;
use Mojo::UserAgent;

my $driver = "SQLite";
my $dsn = "DBI:$driver:dbname=/tmp/misconfiguration.db";
my $dbh = DBI->connect($dsn, undef, undef, { sqlite_see_if_its_a_number => 1, AutoCommit => 1 }) or exit;

my $username = "";
my $userTokenMapping = "";
my $app = Mojolicious->new;
$app = $app->log(Mojo::Log->new(path => '/dev/null'));


######################   API   ##############################

$app->hook(before_dispatch => sub {
    #
    #    Get token from HTTP header
    #
    my $c = shift;
    my $endpoint = $c->req->url->path;

    if ($endpoint eq '/get-token'){
        return;
    }

    my $token = $c->req->headers->header('X-API-TOKEN');
    my $debug = 1;
    if ($token eq ''){
     if ($debug){
      my $sth = $dbh->prepare("SELECT username FROM users");
      $sth->execute;
      my $userMapping = "";
      my $ary =  $sth->fetchall_arrayref({});
      for my $i ( 0 .. $#$ary ) {
          $userMapping .= $ary->[$i]{'username'} . ":";
      }
      return $c->render(
          status => 401,
          template => 'unauthorized',
          content => 'DEBUG: username list: '.$userMapping
      );
    }
    return $c->render(
           status => 401,
           template => 'unauthorized',
           content => 'Missing or invalid token'
    );
    }


    my $sth = $dbh->prepare("SELECT username FROM users WHERE token=?");
    $sth->bind_param(1, $token);
    $sth->execute;
    $userTokenMapping = $sth->fetchrow_array;

    if (!$userTokenMapping) {
      if ($debug){
        my $sth = $dbh->prepare("SELECT token FROM users");
        $sth->execute;
        my $tokenMapping = "";
  	my $ary =  $sth->fetchall_arrayref({});
        	for my $i ( 0 .. $#$ary ) {
            $tokenMapping .= $ary->[$i]{'token'} . ":";
        	}
          return $c->render(
              status => 401,
              template => 'unauthorized',
              content => 'DEBUG: token list: ' . $tokenMapping
          );
        }
        return $c->render(
              status => 401,
              template => 'unauthorized',
              content => 'Invalid X-Api-Token'
        );
      }

});


$app->routes->get('/me' => sub {
    #
    #    Returns username based on api token value
    #
    my $c = shift;
    return $c->render(
        status => 200,
        text => "Your token match the user '$userTokenMapping'"
    );
});


$app->routes->get('/get-secret' => sub {
    #
    #    Return secret page
    #
    my $c = shift;
    if ($userTokenMapping ne 'admin') {
        return $c->render(
            status => 403,
            template => 'forbidden',
            content => "You are not allowed to request /get-secret"
        );
    }

    open my $fh, '<', 'secret' or die "error opening secret file";
    my $secret = do { local $/; <$fh> };
    return $c->render(
        status => 200,
        text => "You are logged in. And congratz ! Here is the secret : $secret"
    );
});

####################     Main   #############################

sub main {

    $username = $ARGV[0];
    my $token = $ARGV[1];
    my $endpoint = $ARGV[2];

    if (!$endpoint) {
        $endpoint = '/me';
    }

    my $url = Mojo::URL->new($endpoint);
    my $ua = Mojo::UserAgent->new();
    $ua->server->app($app);
    my $response = $ua->get($url => {'X-API-TOKEN' => $token});
    print $response->res->content->asset->{'content'};
    $dbh->disconnect();
}

main();
1;

__DATA__

@@ forbidden.html.ep
% layout 'default';
<h1>Forbidden</h1>
<p><%= $content %></p>

@@ unauthorized.html.ep
% layout 'default';
<h1>Unauthorized</h1>
<p><%= $content %></p>

@@ not_found.html.ep
% layout 'default';
<h1>Not found</h1>

@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head><title>MyApp</title></head>
  <body><%= content %></body>
</html>

File: challs/misconfiguration.dir/misconfiguration.py
#!/usr/bin/python3

import sqlite3
import sys

from flask import Flask, g, request
from werkzeug.exceptions import (BadRequest, Forbidden,
                                 Unauthorized)


def get_params():
    """
        Check script params
    """
    if len(sys.argv) != 4:
        print("Missing parameters")
        sys.exit(0)

    username = sys.argv[1]
    token = sys.argv[2]
    endpoint = sys.argv[3]

    return username, token, endpoint


def create_app():
    """
        Initialize Flask application
    """
    app = Flask(__name__)

    @app.before_request
    def get_user_from_token():
        """
            Get token from HTTP header
        """

        try:
            token = request.environ['HTTP_X_API_TOKEN']
        except (KeyError, IndexError, TypeError):
            if APP.config['DEBUG']:
                g.cursor.execute("SELECT username FROM users")
                rows = g.cursor.fetchall()
                username = ""
                for row in rows:
                    username = row[0] + "\n" + username
                raise BadRequest("DEBUG: username list: {}".format(username))
            raise BadRequest('Missing or invalid token')

        g.cursor.execute("SELECT username FROM users WHERE token=?", (token,))
        user = g.cursor.fetchone()
        if not user:
            if APP.config['DEBUG']:
                g.cursor.execute("SELECT token FROM users")
                rows = g.cursor.fetchall()
                token = ""
                for row in rows:
                    token = row[0] + "\n" + token
                raise BadRequest("DEBUG: token list: {}".format(token))
            raise Unauthorized("Invalid X-Api-Token")

        g.user_token_mapping = user[0]  # Store username in the request context

    @app.route('/me')
    def me():
        """
            Returns username based on api token value
        """
        return "Your token match the user '{}'".format(g.user_token_mapping)

    @app.route('/get-secret')
    def get_secret():
        """
            Return secret page
        """
        if g.user_token_mapping != 'admin':
            raise Forbidden('You are not allowed to request /get-secret')

        with open('secret') as secret:
            return "You are logged in. And congratz ! Here is the secret : {}".format(
                secret.read()
            )

    return app


APP = create_app()
# please remove this setting in production.
APP.config['DEBUG'] = True
APP.config['TESTING'] = True


if __name__ == '__main__':

    # Parse params
    username, token, endpoint = get_params()

    # Init app and db cursor
    try:
        tester = APP.test_client()
        ctx = APP.test_request_context()
        ctx.push()
        conn = sqlite3.connect('/tmp/misconfiguration.db', isolation_level=None)
        cursor = conn.cursor()
        g.cursor = cursor
        g.username = username
    except:
        print('Error while connecting to db.')
        sys.exit(1)

    if not endpoint:
        endpoint = '/me'

    if token:
        headers = {'X-API-TOKEN': token}
    else:
        headers = ""

    # Make request
    response = tester.get(
        endpoint,
        headers=headers,
    )

    conn.close()
    print(response.get_data().decode('unicode_escape'))
